### ISSUE TITLE
Infinite update loop on FederatedDeployment

### ISSUE THREAD

**Issue by rudoi** [kubernetes-sigs/kubefed#982]
**What happened**:

I federated a Deployment:
`kubefedctl federate -f file/with/deployment.yaml | kubectl apply -f`

It seems to be updating forever:
```
Events:
  Type     Reason                 Age                   From                            Message
  ----     ------                 ----                  ----                            -------
  Normal   CreateInCluster        31m                   federateddeployment-controller  Creating Deployment "default/dummy" in cluster "ar2a"
  Warning  UpdateInClusterFailed  30m (x4 over 31m)     federateddeployment-controller  Failed to update Deployment "default/dummy" in cluster "ar2a": Operation cannot be fulfilled on deployments.apps "dummy": the object has been modified; please apply your changes to the latest version and try again
  Normal   UpdateInCluster        6m4s (x655 over 31m)  federateddeployment-controller  Updating Deployment "default/dummy" in cluster "ar2a"
  Normal   UpdateInCluster        48s (x15 over 69s)    federateddeployment-controller  Updating Deployment "default/dummy" in cluster "ar2a"
  Warning  UpdateInClusterFailed  48s (x4 over 69s)     federateddeployment-controller  Failed to update Deployment "default/dummy" in cluster "ar2a": Operation cannot be fulfilled on deployments.apps "dummy": the object has been modified; please apply your changes to the latest version and try again
  Normal   UpdateInCluster        2s (x4 over 12s)      federateddeployment-controller  Updating Deployment "default/dummy" in cluster "ar2a"
```

**What you expected to happen**:

Reconciliation to finish and the status to be updated.

**How to reproduce it (as minimally and precisely as possible)**:

Deploy a FederatedDeployment.

**Environment**:
- Kubernetes version (use `kubectl version`)
```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.0", GitCommit:"641856db18352033a0d96dbc99153fa3b27298e5", GitTreeState:"clean", BuildDate:"2019-03-26T00:04:52Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.7", GitCommit:"4683545293d792934a7a7e12f2cc47d20b2dd01b", GitTreeState:"clean", BuildDate:"2019-06-06T01:39:30Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```
- KubeFed version: 0.1.0-rc2
- Scope of installation (namespaced or cluster): cluster
- Others

/kind bug

**Comment by marun** [comment]
As per @zeronone's discovery in #1006, the reason for the thrashing appears to be that the deployment controller in the member cluster is setting an annotation that the sync controller is clearing. The near-term fix is likely to be retaining the specific annotation.  A longer-term fix may be supporting path-based retention.

**Comment by zeronone** [comment]
Path-based retention would be nice, but the user has to manually enable them for each case (needs knowledge of each case). Would changing the semantics for `metata.annotations`, `metadata.finalizers` be a bad idea? i.e: reconciliation always do a "merge" on those properties rather than overwriting them?

I am currently using Rancher, and I can't use `FederatedNamespace` either. Since rancher puts some annotations, finalizers on the namespace object (for user authorization purposes).

**Comment by marun** [comment]
Merging might work for addition, but what if a user has added an annotation and then wants to remove it? Handling this effectively would require deviating from the edge-based reconciliation strategy espoused by Kubernetes (i.e. tracking intermediate states).

We've seen a number of issues related to this problem of kubefed reconciliation conflicting with local controllers, though. I think your intuition is correct that the `finalizers` and `annotations` fields need special treatment.  In the case of `finalizers`, I think there is a good case to be made for the sync controller always retaining the field of an existing resource since the field should only ever be set by controllers local to a cluster.

`annotations` would seem to require more specialized behavior. The near-term solution could be manually enabling retention as we discover a need to do so (e.g. deployments, rancher, etc), or simply retaining all annotations.

My revised thinking on a longer-term solution would be for the sync controller to always retain the `annotations` field and for changes to the field to be declared via overrides. Once support for jsonpatch overrides has merged (#991), we can add support for the `add` and `remove` operations that to support changing a potentially populated collection field like `annotations`.  It would probably be desirable to also implement selector-based overrides (#581) to simplify the definition of an override that should apply to multiple clusters (e.g. set an annotation on the resource in all clusters.

As to your last question, the orphan annotation (not label) is only applied to the federated resource rather than the template, so it should not be affected by this issue.

cc: @shashidharatd

**Comment by marun** [comment]
@rudoi #1011 should resolve the problem you are seeing

@zeronone #1010 should resolve the finalizer issue you are seeing, and #1011 should enable retaining the annotations that rancher is setting on namespaces. You will need to create a PR based on #1011 and add the keys of the rancher annotations so they can be retained.
